spring-shell-core/src/main/java/org/springframework/shell/Shell.java

```diff
@@ -264,4 +284,24 @@ private String findLongestCommand(String prefix) {
 		return "".equals(result) ? null : result;
 	}
 
+	private static class ParameterToResolver implements Comparable<ParameterToResolver> {
```

Comment:
Let's assume we have 2 parameters, e.g. `methodName(T1 arg1, T2 arg2)`, and 3 parameter resolvers `[P1, P2 and P3]` (in priority order) and the possible resolvers for `arg1` -> `P1` and `P3`, and for `arg2` -> `P2` and `P3`. During a `resolve` operation, my understanding is that the operation should attempt the following in order:

1. try to resolve `arg1` using `P1`
2. try to resolve `arg2` using `P2`
3. try to resolve `arg1` using `P3` (in case `P1` failed to resolve)
4. try to resolve `arg2` using `P3` (in case `P2` failed to resolve)

In 1 and 3 are why we need to return multiple resolvers (and not the first that matches) as we'd need to try the next compatible resolver (respecting the priority of the other resolvers).

In 3 and 4 are why the index comes into place, the same resolver should be invoked in parameter order.

The reason for the tuple is to pre-compute the order the operation will use.

In a more concrete example, the way I see this is what we discussed last time as a possible approach, for example, StandardParameterResolver being potentially broken down into:

1. `P1` -> StandardNamedParameterResolver
2. `P2` -> StandardBooleanParameterResolver (I don't actually had this in mind, but works with the example above :) - but the boolean 0-arity case could actually be handled like this, again, just for the example, but worth thinking about)
3. `P3` -> StandardPositionalParameterResolver

Let me know if you had a different idea in mind, let's try to discuss tomorrow, but will let you know if I need to cancel (in case of family news :) )